core/kernel/user_access.c

```diff
@@ -74,6 +96,137 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
 	return res;
 }
 
+void *bb_alloc(size_t len)
+{
+	struct user_mode_ctx *uctx = get_current_uctx();
+	size_t offs = 0;
+	void *bb = NULL;
+
+	if (uctx && !ADD_OVERFLOW(uctx->bbuf_offs, len, &offs) &&
+	    offs <= uctx->bbuf_size) {
+		bb = uctx->bbuf + uctx->bbuf_offs;
+		uctx->bbuf_offs = ROUNDUP(offs, BB_ALIGNMENT);
+	}
+	return bb;
+}
+
+static void bb_free_helper(struct user_mode_ctx *uctx, vaddr_t bb, size_t len)
+{
+	vaddr_t bbuf = (vaddr_t)uctx->bbuf;
+
+	if (bb > bbuf && IS_ALIGNED(bb, BB_ALIGNMENT)) {
+		size_t prev_offs = bb - bbuf;
+
+		if (prev_offs + ROUNDUP(len, BB_ALIGNMENT) == uctx->bbuf_offs)
+			uctx->bbuf_offs = prev_offs;
+	}
+}
+
+void bb_free(void *bb, size_t len)
+{
+	struct user_mode_ctx *uctx = get_current_uctx();
+
+	if (uctx)
+		bb_free_helper(uctx, (vaddr_t)bb, len);
+}
+
+void bb_reset(void)
+{
+	struct user_mode_ctx *uctx = get_current_uctx();
+
+	if (uctx)
+		uctx->bbuf_offs = 0;
+}
+
+TEE_Result clear_user(void *uaddr, size_t n)
+{
+	uint32_t flags = TEE_MEMORY_ACCESS_WRITE | TEE_MEMORY_ACCESS_ANY_OWNER;
+	TEE_Result res = TEE_SUCCESS;
+
+	uaddr = memtag_strip_tag(uaddr);
+	res = check_access(flags, uaddr, n);
+	if (res)
+		return res;
+
+	memset(uaddr, 0, n);
+
+	return TEE_SUCCESS;
+}
+
+size_t strnlen_user(const void *uaddr, size_t len)
+{
+	uint32_t flags = TEE_MEMORY_ACCESS_READ | TEE_MEMORY_ACCESS_ANY_OWNER;
+	TEE_Result res = TEE_SUCCESS;
+	size_t n = 0;
+
+	uaddr = memtag_strip_tag_const(uaddr);
+	res = check_access(flags, uaddr, len);
+	if (!res)
+		n = strnlen(uaddr, len);
+
+	return n;
+}
+
+TEE_Result bb_memdup_user(const void *src, size_t len, void **p)
+{
+	TEE_Result res = TEE_SUCCESS;
+	void *buf = NULL;
+
+	buf = bb_alloc(len);
+	if (!buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = copy_from_user(buf, src, len);
+	if (res)
+		bb_free(buf, len);
+	else
+		*p = buf;
+
+	return res;
+}
+
+TEE_Result bb_memdup_user_private(const void *src, size_t len, void **p)
+{
+	TEE_Result res = TEE_SUCCESS;
+	void *buf = NULL;
+
+	buf = bb_alloc(len);
+	if (!buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = copy_from_user_private(buf, src, len);
+	if (res)
+		bb_free(buf, len);
+	else
+		*p = buf;
+
+	return res;
+}
+
+TEE_Result bb_strndup_user(const char *src, size_t maxlen, char **dst,
+			   size_t *dstlen)
+{
+	uint32_t flags = TEE_MEMORY_ACCESS_READ | TEE_MEMORY_ACCESS_ANY_OWNER;
+	TEE_Result res = TEE_SUCCESS;
+	size_t l = 0;
+	char *d = NULL;
+
+	src = memtag_strip_tag_const(src);
+	res = check_access(flags, src, maxlen);
+	if (res)
+		return res;
+	l = strnlen(src, maxlen);
```

Comment:
how about using `strnlen_user()`? Otherwise, it would raise PAN error.

Reply:
I'd expect that `bb_strndup_user()` should also be updated to use `enter_user_access()` and `exit_user_access()`.